src/number_theory/pell.lean

```diff
@@ -144,6 +144,141 @@ lemma x_neg (a : solution₁ d) : (-a).x = -a.x := rfl
+begin
+  simp only [y_mul],
+  positivity,
+end
+
+/-- If `(x, y)` is a solution with `x` positive, then all its powers with natural exponents
+have positive `x`. -/
+lemma x_pow_pos {a : solution₁ d} (hax : 0 < a.x) (n : ℕ) : 0 < (a ^ n).x :=
+begin
+  induction n with n ih,
+  { simp only [pow_zero, x_one, zero_lt_one], },
+  { rw [pow_succ],
+    exact x_mul_pos hax ih, }
+end
+
+/-- If `(x, y)` is a solution with `x` and `y` positive, then all its powers with positive
+natural exponents have positive `y`. -/
+lemma y_pow_succ_pos {a : solution₁ d} (hax : 0 < a.x) (hay : 0 < a.y) (n : ℕ) :
+  0 < (a ^ n.succ).y :=
+begin
+  induction n with n ih,
+  { simp only [hay, pow_one], },
+  { rw [pow_succ],
+    exact y_mul_pos hax hay (x_pow_pos hax _) ih, }
+end
+
+/-- If `(x, y)` is a solution with `x` positive, then all its powers have positive `x`. -/
+lemma x_zpow_pos {a : solution₁ d} (hax : 0 < a.x) (n : ℤ) : 0 < (a ^ n).x :=
+begin
+  cases le_or_lt 0 n with h h,
+  { rw [← int.to_nat_of_nonneg h, zpow_coe_nat],
+    exact x_pow_pos hax _, },
+  { rw [← neg_neg n, zpow_neg, x_inv, ← int.to_nat_of_nonneg (neg_nonneg.mpr h.le), zpow_coe_nat],
+    exact x_pow_pos hax _, },
+end
+
+/-- If `(x, y)` is a solution with `x` and `y` positive, then the `y` component of any power
+has the same sign as the exponent. -/
+lemma sign_y_zpow_eq_sign_of_x_pos_of_y_pos {a : solution₁ d} (hax : 0 < a.x) (hay : 0 < a.y)
+   (n : ℤ) :
+  (a ^ n).y.sign = n.sign :=
+begin
+  have H : ∀ m : ℤ, 0 < m → 0 < (a ^ m).y,
+  { change ∀ m : ℤ, 1 ≤ m → _,
+    refine λ m, int.le_induction (by simp only [hay, zpow_one]) (λ m hm ih, _) m,
+    rw zpow_add_one,
+    exact y_mul_pos (x_zpow_pos hax m) ih hax hay, },
+  rcases lt_trichotomy 0 n with hpos | rfl | hneg,
+  { rw [(int.sign_eq_one_iff_pos n).mpr hpos, int.sign_eq_one_iff_pos],
+    exact H n hpos, },
+  { rw [int.sign_zero, int.sign_eq_zero_iff_zero, zpow_zero, y_one], },
+  { rw [(int.sign_eq_neg_one_iff_neg n).mpr hneg, int.sign_eq_neg_one_iff_neg, ← neg_neg n,
+        zpow_neg, y_inv, neg_lt, neg_zero],
+    exact H (-n) (lt_neg.mp hneg), }
+end
+
+/-- If `a` is any solution, then one of `a`, `a⁻¹`, `-a`, `-a⁻¹` has
+positive `x` and nonnegative `y`. -/
+lemma exists_pos_variant (h₀ : 0 < d) (a : solution₁ d) :
+  ∃ b : solution₁ d, 0 < b.x ∧ 0 ≤ b.y ∧ a ∈ ({b, b⁻¹, -b, -b⁻¹} : set (solution₁ d)) :=
+begin
+  refine ⟨mk (|a.x|) (|a.y|) (by simp [a.prop]),
+          abs_pos.mpr (a.x_ne_zero h₀.le), abs_nonneg a.y, _⟩,
+  cases le_or_lt 0 a.x with hax hax; cases le_or_lt 0 a.y with hay hay,
+  { exact or.inl (ext (abs_of_nonneg hax).symm (abs_of_nonneg hay).symm), },
+  { exact or.inr (or.inl $ ext (by simp [abs_of_nonneg hax]) (by simp [abs_of_neg hay])), },
+  { exact or.inr (or.inr $ or.inr $ ext (by simp [abs_of_neg hax])
+                                        (by simp [abs_of_nonneg hay])), },
+  { exact or.inr (or.inr $ or.inl $ ext (by simp [abs_of_neg hax]) (by simp [abs_of_neg hay])),}
```

Comment:
```suggestion
  refine (lt_or_gt_of_ne (a.x_ne_zero h₀)).elim ((le_total 0 a.y).elim (λ hy hx, ⟨-a⁻¹, _, _, _⟩)
    (λ hy hx, ⟨-a, _, _, _⟩)) ((le_total 0 a.y).elim (λ hy hx, ⟨a, hx, hy, _⟩)
    (λ hy hx, ⟨a⁻¹, hx, _, _⟩)); simp; assumption,
```
Here's the squeezed simp: `simp only [neg_neg, inv_inv, neg_inv, set.mem_insert_iff, set.mem_singleton_iff,
      eq_self_iff_true, x_neg, x_inv, y_neg, y_inv, neg_pos, neg_nonneg, or_true, true_or]`